llvm/lib/Target/BPF/BPFISelLowering.cpp

```diff
@@ -481,8 +495,47 @@ SDValue BPFTargetLowering::LowerFormalArguments(
 
   if (Subtarget->isSolana()) {
     if (IsVarArg) {
-      fail(DL, DAG, "Functions with VarArgs are not supported");
-      assert(false);
+      ArrayRef<MCPhysReg> ArgRegs = makeArrayRef(SBFArgRegs);
+      unsigned Idx = CCInfo.getFirstUnallocated(ArgRegs);
+      const TargetRegisterClass *RC = getRegClassFor(MVT::i64);
+
+      // Offset of the first variable argument from stack pointer.
+      int VaArgOffset;
+
+      if (ArgRegs.size() == Idx)
+        VaArgOffset = alignTo(CCInfo.getNextStackOffset(), 8);
+      else {
+        VaArgOffset = -((int)(8 * (ArgRegs.size() - Idx)));
+      }
+
+      // Record the frame index of the first variable argument
+      // which is a value necessary to VASTART.
+      int FI = MFI.CreateFixedObject(8, VaArgOffset, true);
+      BFI->setVarArgsFrameIndex(FI);
+
+      // Used with vargs to acumulate store chains.
+      std::vector<SDValue> OutChains;
+
+      for (unsigned I = Idx; I < ArgRegs.size();
+          ++I, VaArgOffset += 8) {
+        Register Reg = MF.getRegInfo().createVirtualRegister(RC);
+        MF.getRegInfo().addLiveIn(ArgRegs[I], Reg);
+        SDValue ArgValue = DAG.getCopyFromReg(Chain, DL, Reg, MVT::i64);
+        FI = MFI.CreateFixedObject(8, VaArgOffset, true);
+        SDValue PtrOff = DAG.getFrameIndex(FI, getPointerTy(DAG.getDataLayout()));
+        SDValue Store =
+            DAG.getStore(Chain, DL, ArgValue, PtrOff, MachinePointerInfo::getFixedStack(MF, FI));
+        cast<StoreSDNode>(Store.getNode())->getMemOperand()->setValue(
+            (Value *)nullptr);
+        OutChains.push_back(Store);
```

Comment:
This store is too strong. Clang with `-O3` cannot eliminate stores of unused varargs backups.

```
0000000000000000 <varargs>:
       0:       17 0b 00 00 20 00 00 00 r11 -= 32
       1:       bf 20 00 00 00 00 00 00 r0 = r2
       2:       7b 4a f8 ff 00 00 00 00 *(u64 *)(r10 - 8) = r4
       3:       7b 3a f0 ff 00 00 00 00 *(u64 *)(r10 - 16) = r3
       4:       7b 0a e8 ff 00 00 00 00 *(u64 *)(r10 - 24) = r0
       5:       0f 10 00 00 00 00 00 00 r0 += r1
       6:       07 0b 00 00 20 00 00 00 r11 += 32
       7:       95 00 00 00 00 00 00 00 exit
```